File: c++/guess-number.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <limits>

using namespace std;

void printWelcome() {
    cout << "========================================" << endl;
    cout << "     欢迎来到猜数字游戏！" << endl;
    cout << "========================================" << endl;
    cout << "规则说明：" << endl;
    cout << "1. 系统会随机生成一个1-100之间的整数" << endl;
    cout << "2. 你有10次机会猜出这个数字" << endl;
    cout << "3. 每次猜测后，系统会提示猜大或猜小" << endl;
    cout << "========================================" << endl << endl;
}

void printResult(int guess, int target, int remaining) {
    if (guess > target) {
        cout << "猜大了！再小一点。" << endl;
    } else if (guess < target) {
        cout << "猜小了！再大一点。" << endl;
    }
    cout << "剩余机会: " << remaining << " 次" << endl << endl;
}

int getPlayerGuess() {
    int guess;
    while (true) {
        cout << "请输入你的猜测 (1-100): ";
        cin >> guess;
        
        // 检查输入是否有效
        if (cin.fail()) {
            cin.clear(); // 清除错误状态
            cin.ignore(numeric_limits<streamsize>::max(), '\n'); // 忽略错误输入
            cout << "输入无效！请输入一个整数。" << endl;
            continue;
        }
        
        // 检查范围
        if (guess < 1 || guess > 100) {
            cout << "数字必须在 1-100 之间！" << endl;
            continue;
        }
        
        break;
    }
    return guess;
}

bool playAgain() {
    char choice;
    while (true) {
        cout << "是否再来一局？(y/n): ";
        cin >> choice;
        if (choice == 'y' || choice == 'Y') {
            return true;
        } else if (choice == 'n' || choice == 'N') {
            return false;
        } else {
            cout << "请输入 y 或 n" << endl;
        }
    }
}

int main() {
    srand(static_cast<unsigned int>(time(nullptr))); // 初始化随机种子
    
    do {
        printWelcome();
        
        int target = rand() % 100 + 1; // 生成1-100的随机数
        int maxAttempts = 10;
        int attempts = 0;
        bool guessed = false;
        
        cout << "游戏开始！我已选好了一个1-100之间的数字。" << endl;
        cout << "你有 " << maxAttempts << " 次机会，祝你好运！" << endl << endl;
        
        while (attempts < maxAttempts) {
            int remaining = maxAttempts - attempts;
            cout << "--- 第 " << (attempts + 1) << " 次猜测 ---" << endl;
            
            int guess = getPlayerGuess();
            attempts++;
            remaining--;
            
            if (guess == target) {
                cout << "恭喜你猜对了！就是 " << target << " ！" << endl;
                cout << "你用了 " << attempts << " 次机会。" << endl;
                guessed = true;
                break;
            } else {
                printResult(guess, target, remaining);
            }
        }
        
        if (!guessed) {
            cout << "很遗憾，10次机会用完了！" << endl;
            cout << "正确答案是: " << target << endl;
        }
        
        cout << endl;
        
    } while (playAgain());
    
    cout << "感谢游玩！" << endl;
    return 0;
}

File: go/guess-number.go
package main

import (
    "fmt"
	"math/rand"
 	"time"
)

func main() {
	// 使用当前时间作为随机种子，确保每次运行数字不同
	rand.Seed(time.Now().UnixNano())
	// 生成1-100之间的随机数
	target := rand.Intn(100) + 1
	
	var guess int
	maxAttempts := 10
	attempts := 0
	
	fmt.Println("====================================")
	fmt.Println("     欢迎来到猜数字游戏！")
	fmt.Println("====================================")
	fmt.Printf("游戏规则：")
	fmt.Printf("1. 系统会随机生成一个1-100之间的数字")
	fmt.Printf("2. 你有10次机会猜中这个数字")
	fmt.Printf("3. 每次猜测后会提示'大了'或'小了'")
	fmt.Printf("你有 %d 次机会来猜中它。\n\n", maxAttempts)
	
	// 主循环，最多尝试 maxAttempts 次
	for attempts < maxAttempts {
		fmt.Printf("请输入你的猜测 (剩余 %d 次): ", maxAttempts-attempts)
		_, err := fmt.Scan(&guess)
		if err != nil {
			fmt.Println("输入无效，请输入一个整数。")
			// 清除输入缓冲区中的错误数据
			var discard string
			fmt.Scanln(&discard)
			continue
		}
		
		// 检查输入范围
		if guess < 1 || guess > 100 {
			fmt.Println("   请输入 1 到 100 之间的数字。")
			continue
		}
		
		attempts++
		
		// 比较猜测与目标值
		if guess == target {
			fmt.Printf("\n恭喜你！你猜对了！数字就是 %d。\n", target)
			fmt.Printf("你用了 %d 次机会。\n", attempts)
			return
		} else if guess < target {
			fmt.Println("太小了，再大一点。")
		} else {
			fmt.Println("太大了，再小一点。")
		}
		
		// 在最后一次机会时给出提示
		if attempts == maxAttempts {
			fmt.Printf("\n很遗憾，你的 %d 次机会已经用完。\n", maxAttempts)
			fmt.Printf("正确的数字是 %d。\n", target)
			fmt.Println("下次加油！")
		}
	}
}